src/toil/lib/aws/utils.py

```diff
@@ -0,0 +1,96 @@
+
+logger = logging.getLogger(__name__)
+
+
+@retry(errors=[BotoServerError])
+def delete_iam_role(role_name: str, region: Optional[str] = None, quiet: bool = True):
+    from boto.iam.connection import IAMConnection
+    iam_client = aws.client('iam', region_name=region)
+    iam_resource = aws.resource('iam', region_name=region)
+    boto_iam_connection = IAMConnection()
+    role = iam_resource.Role(role_name)
+    # normal policies
+    for attached_policy in role.attached_policies.all():
+        printq(f'Now dissociating policy: {attached_policy.name} from role {role.name}', quiet)
+        role.detach_policy(PolicyName=attached_policy.name)
+    # inline policies
+    for attached_policy in role.policies.all():
+        printq(f'Deleting inline policy: {attached_policy.name} from role {role.name}', quiet)
+        # couldn't find an easy way to remove inline policies with boto3; use boto
+        boto_iam_connection.delete_role_policy(role.name, attached_policy.name)
+    iam_client.delete_role(RoleName=role_name)
+    printq(f'Role {role_name} successfully deleted.', quiet)
+
+
+@retry(errors=[BotoServerError])
+def delete_iam_instance_profile(instance_profile_name: str, region: Optional[str] = None, quiet: bool = True):
+    iam_resource = aws.resource('iam', region_name=region)
+    instance_profile = iam_resource.InstanceProfile(instance_profile_name)
+    for role in instance_profile.roles:
+        printq(f'Now dissociating role: {role.name} from instance profile {instance_profile_name}', quiet)
+        instance_profile.remove_role(RoleName=role.name)
+    instance_profile.delete()
+    printq(f'Instance profile "{instance_profile_name}" successfully deleted.', quiet)
+
+
+@retry(errors=[BotoServerError])
+def delete_sdb_domain(sdb_domain_name: str, region: Optional[str] = None, quiet: bool = True):
+    sdb_client = aws.client('sdb', region_name=region)
+    sdb_client.delete_domain(DomainName=sdb_domain_name)
+    printq(f'SBD Domain: "{sdb_domain_name}" successfully deleted.', quiet)
+
+
+@retry(errors=[BotoServerError])
+def delete_s3_bucket(bucket: str, region: Optional[str], quiet: bool = True):
+    printq(f'Deleting s3 bucket in region "{region}": {bucket}', quiet)
+    s3_client = aws.client('s3', region_name=region)
+    s3_resource = aws.resource('s3', region_name=region)
+
+    paginator = s3_client.get_paginator('list_object_versions')
+    for response in paginator.paginate(Bucket=bucket):
+        versions = response.get('Versions', []) + response.get('DeleteMarkers', [])
+        for version in versions:
+            printq(f"    Deleting {version['Key']} version {version['VersionId']}", quiet)
+            s3_client.delete_object(Bucket=bucket, Key=version['Key'], VersionId=version['VersionId'])
+    s3_resource.Bucket(bucket).delete()
+    printq(f'\n * Deleted s3 bucket successfully: {bucket}\n\n', quiet)
+
+
+def bucket_in_use():
+    pass
+
+
+def iam_instance_profile_in_use():
+    pass
```

Comment:
Did you forget about these or are they meant to be empty? What are they supposed to do?

Reply:
I just haven't taken the time to fill them in yet.  I wanted to implement them after your last comment and just put in a reminder for myself.